core/assay.js

```diff
@@ -0,0 +1,131 @@
+import harden from '@agoric/harden';
+
+import { insist } from '../util/insist';
+import { mustBeSameStructure, mustBeComparable } from '../util/sameStructure';
+
+// This module treats labels as black boxes. It is not aware
+// of issuers, and so can handle labels whose issuers are merely
+// presences of remote issuers.
+
+// `makeAssayMaker` takes in a 'strategy' object which defines set
+// operations for a particular kind of assay. For instance, for the
+// natAssay, these operations are arithmetic.
+
+// Return an assay, which makes amounts, validates amounts, and
+// provides set operations over amounts. An amount is a pass-by-copy
+// description of some set of erights. An amount has a label and a
+// quantity. All amounts made by the same assay have the same label
+// but differ in quantity.
+//
+// An assay is pass-by-presence, but is not designed to be usefully
+// passed. Rather, we expect each vat that needs to operate on amounts
+// will have its own local assay to do so.
+
+function makeAssayMaker(strategy) {
+  function makeAssay(label) {
+    mustBeComparable(label);
+
+    // The brand represents recognition of the amount as authorized.
+    const brand = new WeakSet();
+
+    const assay = harden({
+      getLabel() {
+        return label;
+      },
+
+      getStrategy() {
+        return strategy;
+      },
+
+      // Given the raw quantity that this kind of amount would label, return
+      // an amount so labeling that quantity.
+      make(allegedQuantity) {
+        const amount = harden({
+          label,
+          quantity: strategy.insistKind(allegedQuantity),
+        });
+        brand.add(amount);
+        return amount;
+      },
+
+      // Is this an amount object made by this assay? If so, return
+      // it. Otherwise error.
+      vouch(amount) {
+        insist(brand.has(amount))`\
+  Unrecognized amount: ${amount}`;
+        return amount;
+      },
+
+      // Is this like an amount object made by this assay, such as one
+      // received by pass-by-copy from an otherwise-identical remote
+      // amount? On success, return an amount object made by this
+      // assay. Otherwise error.
+      //
+      coerce(allegedAmount) {
+        if (brand.has(allegedAmount)) {
+          return allegedAmount;
+        }
+        if (!Object.prototype.hasOwnProperty.call(allegedAmount, 'quantity')) {
+          // This is not an amount. Let's see if it's a quantity. Will
```

Comment:
This logic comes from the NatAssay, where we wanted users to be able to mint new tokens easily, such as calling `moolaMint.mint(1000)`. Without it, we would be forcing users to have to create an amount to pass into the `mint` method. Allowing users to pass in a quantity when minting makes the other kinds of mints easier to use as well. For instance, in the pixel gallery, the gallery currently does:

```
const amount = galleryPixelAssay.make(harden([pixel]));
const purse = galleryPixelMint.mint(amount);
```
This could be instead: 
```
const purse = galleryPixelMint.mint(harden[pixel]);
```
You pointed out that a user could be mistaken about what kind of mint they have, and try to mint type1 tokens but actually mint type2 tokens. This is a danger, but I think requiring the user to `make` the amount doesn't add much in solving this problem. By requiring the user to `make` the amount, all we are really checking is that the mint we hold matches the assay that we hold. If I am confused about my assay in the same way as I'm confused about my mint, it won't help me. We're not actually checking that the mint identification is correct. If we trust the user to keep track of what their assays mean, shouldn't we also trust them to keep track of which mints are which? Granted, it is a small double-check to make sure that your mint is the one associated with your assay, but it's a tradeoff and I'm not sure that it is helpful enough to make it more painful for the user to mint new tokens.

Reply:
In your example

> `const purse = galleryPixelMint.mint(harden[pixel]);`

First, did you mean

> `const purse = galleryPixelMint.mint(harden([pixel]));`

Second, what kind of thing is `pixel`?

Reply:
Nevermind. I was misreading your code and intent! I missed the `!` in the condition and thought you were doing something completely different whose motivation I could see but didn't like. Please ignore my comments above.

What if it has a 'quantity' field but no 'label' field? This kind of duck type sniffing is inherently dangerous. But now I understand the real motivation and I do not object. Resolving.